### ISSUE TITLE
v1.0.5 breaks hooks that update the Entry

### ISSUE THREAD

**Issue by fasaxc** [sirupsen/logrus#729]
We've been using a Hook to add line number and file name information to the Entry for a long time.  Unfortunately https://github.com/sirupsen/logrus/pull/707 has broken our Hook.  Since it copies the Entry, our updates to it are lost when the fireHooks method returns.

Our hook used this pattern to avoid mutating the Entry while it may have been read by another thread:
```go
newData := make(log.Fields, len(entry.Data)+2)
for k, v := range entry.Data {
	newData[k] = v
}
newData["__file__"] = path.Base(frame.File)
newData["__line__"] = frame.Line
entry.Data = newData
```
but that no longer works.  I think that simply mutating the existing data map would be broken too.

This seems to work for a quick fix since it's the copy that gets logged out instead of the original Entry:
```
- entry.fireHooks()
+ entry = entry.fireHooks()
...
- func (entry Entry) fireHooks() {
+ func (entry Entry) fireHooks() Entry {
 	entry.Logger.mu.Lock()
 	defer entry.Logger.mu.Unlock()
 	err := entry.Logger.Hooks.Fire(entry.Level, &entry)
 	if err != nil { 
 		fmt.Fprintf(os.Stderr, "Failed to fire hook: %v\n", err)
 	}
+	return entry
 }
```
I'll put up a PR for that.

**Comment by gguridi** [comment]
Yes, that broke existing functionality and I think allowing hooks to modify entries before other hooks dispatch them is something nice to have as logger.

**Comment by ictar** [comment]
hello， any update about this issue?  It seems that fix from @fasaxc has not been merged yet.  I use v1.0.5 and meet the same issue...

**Comment by dgsb** [comment]
Hello this logrus issue should be fixed on master branch, let me know if everything works as expected.

**Comment by caseydavenport** [comment]
Looks like the fix is working, but not for the formatted equivalents (`Debugf`, `Warnf`, etc.). So, I think there is more to do for this one.

From here: https://github.com/projectcalico/node/pull/208#discussion_r275992402

**Comment by krnowak** [comment]
@caseydavenport : Note that I wasn't testing the latest version, but only 1.0.6 and 1.1.0. Maybe it's fixed in 1.4.1…
